transpile: fixes before re-enabling `--reorganize-definitions`

kkysen opened the pull request:
Most of the changes from #1414 with #1451 split out.  In combination with #1451, supersedes #1414.

fw-immunant commented:
We should probably avoid entirely disabling the caching of zero values; I think I noticed c2rust-transpile taking significantly longer after that change. I believe the rest of these fixes were ready to merge.

kkysen commented:
> We should probably avoid entirely disabling the caching of zero values; I think I noticed c2rust-transpile taking significantly longer after that change. I believe the rest of these fixes were ready to merge.

I removed 828c6974b from this PR for now.  I'll fix that in its own PR, and I also split out the non-enable-by-default parts of `--reorganize-definitions` in #1454 so that it can merge without these fixes.  Is this PR good to go now?

fw-immunant commented:
> > We should probably avoid entirely disabling the caching of zero values; I think I noticed c2rust-transpile taking significantly longer after that change. I believe the rest of these fixes were ready to merge.
> 
> I removed 828c6974b from this PR for now. I'll fix that in its own PR, and I also split out the non-enable-by-default parts of `--reorganize-definitions` in #1454 so that it can merge without these fixes. Is this PR good to go now?

The only hesitation I have about merging this as-is is that the codepaths it alters are not tested by CI, because they only apply when `--reorganize-definitions` is passed to the transpiler.

ahomescu commented on tests/items/src/test_fn_attrs.rs:
```diff
@@ -57,8 +63,7 @@ pub fn test_fn_attrs() {
-        assert!(src.contains(
-            "extern \"C\" {\n    #[link_name = \"inline_extern\"]\n    fn aliased_fn();"
-        ));
+        let aliased_fn_syntax = |public| format!("extern \"C\" {{\n    #[link_name = \"inline_extern\"]\n    {}fn aliased_fn();", public);
```
Could we use raw string literals here (and below) to make this more readable?

kkysen replied:
Yeah.  Fixed in d4b5e979f.  I used raw strings for all of them, not just the ones touched in this diff.
